 <br><font color = blue size=7>Python Programming - Part I of III</font> 

***

**Primer Syllabus:**

Data types and data structures. Input and output.  Flow control and exception handling. Functions and modules. 

Special libraries:
* Introduction to some of the powerful libraries in Python and applications
* Scientific algorithms and manipulations (SciPy);
* Data structures and numerical computing (NumPy);
* Graphics and data visualisation (Matplotlib, seaborn, Pyplot);
* Data analysis (Pandas)

This course is split into numerous parts to keep the notes more manageable and take up less space.


# Introduction

Those of us old enough (and fortunate) to remember, will recall the BBC comedy sketch series – Monty Python’s Flying Circus. Televised during the late 60s and early 70s, it was conceived, written and performed by the comedy geniuses collectively known as Monty Python (or simply The Pythons). These intellectual masters of humour were predominantly Oxbridge graduates. Fast forward two decades, and we see that this comedy series became the inspiration for the name of a new programming language! 

<img src="./img/monty.jpg" style="width:600px">


Python was initially developed by Guido van Rossum in the 1990s, who was a big fan of Monty Python's Flying Circus - so the name has no connection with any large serpent! Although, it's conception dates back to the late 1980s. https://en.wikipedia.org/wiki/History_of_Python


This is a programming course in the use of Python as a scientific computing language for computational finance applications; as part of the Oxford MSc. It is intended for those who are new to  programming in Python. As well as an informal lecture style delivery we will also be doing exercises. We don’t want to be computer scientists or developers. Our goal is solve pricing problems numerically; handle/analyse data and plot data, using Python – simple and enjoyable! Where possible we use as much of the built-in functionality as possible – remember this is not a computer science course. The notes also form part of a textbook which is being prepared for Wiley Finance.

Python is a general-purpose dynamic, interpreted (bytecode-compiled) programming language. This means we can type statements into the interpreter and they are executed immediately. In contrast, languages such as Java and C are compiled languages. 

Python was designed for two reasons - so code written in Python would be easy for humans to read; and to minimize the amount of time required to write code.  

There are no type declarations of variables, parameters, functions, or methods in source code. This makes the code short; well structured; readable and flexible, and one loses the compile-time type checking of the source code. The philosophy is “we code what we think”. Python tracks the types of all values at runtime and flags code that does not make sense as it runs. An excellent way to see how Python code works is to run the Python interpreter and type code right into it. 

Science has traditionally consisted of two major disciplines, theory and experimentation. In the last several decades a third important and exciting component has emerged, i.e. scientific computing. Scientific computing acts as an intersection of the former two areas of science. It is often closely related to theory, but it also has many characteristics in common with experimental work. It is therefore often viewed as a third branch of science.

In most areas of science, computation is an invaluable complement to both experiments and theory. Vast majority of both experimental and theoretical research involve some numerical calculations, simulations or computer modelling. In many studies, pure theory alone is insufficient in validating or demonstrating results. On the other hand, experimentation as a sole means of conducting an investigation may lack the scientific rigour necessary to hold up to scrutiny. Experimental work may not be possible or may prove too costly. 


***

The need for mathematical modelling has never been greater. Most problems based on real life are often too complex and cannot be solved using analytical techniques alone – without computational methods and scientific computing techniques we would be greatly limited to the types of modelling  possible. In finance, unless the pricing problem is fairly basic or ideal, it is highly unlikely that a closed form solution is available.


**Why Python?**


The world is driven by technology with the last decade witnessing the greatest innovation in its advancement. Central to the success of this advancement has been programming languages such as C++, Python and R. This course focuses on Python and its application to finance, health and science.

This is an obvious initial point to address when introducing Python. Python is rapidly becoming the standard in scientific computing, receiving much excitement about the application of Python to mathematical finance (although its presence is being felt in many sectors); its appeal continues to grow in both academia and industry. It is also the fastest growing langauge.

Its features include:


* (very) Simple to pick up language – easy to maintain  
* Python offers concise and readable code
* open-source and free language
* growing amount of add-on modules
* multi-platform (cross platform compatible) - available on Windows, Mac OS, Linux and almost all operating systems (Android also) used by Google, YouTube, Instagram, NASA, CERN, Disney, Dropbox. . .
* particularly easy to interface with C++
* powerful language  can be used extensively to solve problems in finance, economics and healthcare
* Broadly adopted: Popularly used Machine Learning and Data Science


**What is open-source?**
* The programme must be freely distributed
* Source code must be included
* Anyone must be allowed to modify the source code – modification/customisation is in fact encouraged and further distribution is allowed
* Certain licensing requirements need to be fulfilled

There are other requirements that need to be satisfied for open-source approval.


**Where does Python programing fit into your life?**

* 'Super small' so shows up on embedded devices
* Several libraries for building great web apps
* Python popular at Disney and Lucas Film
* Large community of users, easy to find help and documentation
* Has a strong position in scientific computing. 
* Heavily used in science (CERN and NASA) with dedicated libraries to specific areas
* Due to Machine Learning its popularity has taken off like a rocket
* Python is an interpreted language (as opposed to compiled like C/ C++/FORTRAN)


**Python has a library for almost anything**

<img src="./img/Bodleian.jpg" style="width:400px">

Python's strongest point is (probably) its guagillions of wonderfully maintained (and ever-growing) libraries. In programming, a library is a big bundle of code snippets (or functions) that someone else has written and made freely available for other people to download and modify (open source culture again!). A single library will usually be designed to help people write code for specific applications or purposes. For example:

* SciPy is a library for general mathematical purposes
* Pandas is a library for data science
* Numpy is a library for numerical computing
* NLTK is a library for natural language processing
* Tensorflow is a machine learning (ML) library released by Google
* Epipy is a library for epidemiology for analyzing and visualizing epidemiology data
* Matplotlib is a library for graphics (static, animated, interactive)
* Mesa is a library for agent based modelling

Using libraries rather than writing your own code from scratch is always a smart move - you'll save time, and because open source code will have been reviewed and checked by hundreds and thousands of people, code from a library is far less likely to contain bugs than something you may write.


**Why Python for Data Science?**

Data analysis has become one of the fastest developing areas in IT. If you had access to a large dataset, would you be able to find the answers to important questions? We are surrounded and continuously bombarded with data. Data Science has emerged as a highly important academic and industrial based discipline which continues to expand rapidly. Combining both computational and statistical methods for data-driven problem solving, its applications are wide ranging to include finance, health, journalism and politics. This rapidly expanding area includes machine learning, deep learning, large-scale data analysis. So why has Python become so popular in this field?

* Python is an interpreted language (as opposed to compiled e.g. C/ C++/FORTRAN)
* Programmers find it particularly attractive
* Dynamic (i.e. single line of code will run in the Python interpreter)
* Used as glue language, existing legacy scientific computing software (written in C/C++/FORTRAN) can be combined as software libraries in python. 
* As listed above, the availability of many libraries
* Object-oriented (structured around objects that contain variables and methods, where methods act on variables)

**Python is object oriented**

Object-oriented programming, or OOP for short, is a way of structuring programs so that groups of properties and behaviours are bundled into things called 'objects'.

We'll learn more about why this is useful later.

**Why Python for Machine Learning?**
Artificial Intelligence (AI) and Machine Learning (AI) are terms from computer science that are mentioned extensively (whether in passing or at a deeper/technical level) and have become subjects in their own right. While AI has been referred to for a few decades now, ML is relatively new. There is also some confusion as to the difference between the two, and if the names can be used interchangeably. Whilst this discussion can be lengthy, a few key differences can be outlined. 

ML is a branch of AI. AI seeks to simulate natural intelligence to solve complex problems, while the goal of ML is to learn from data on a particular task to optimise the performance of the machine. AI is about decision making, whereas ML allows a system to learn new things by being trained on large amounts of data.

In short ML is the process of teaching a computer system how to make accurate predictions when fed data. It is the technique whereby machines can tackle tasks that up until now have been carried out by humans. Hence it is clear that the key being lots and lots of data!

So why Python? 
* The code is understandable by humans, which makes it easier to build models for ML
* Extensive selection of frameworks and libraries 

>* Keras, TensorFlow, Scikit-learn for machine learning
>* NumPy for high-performance scientific computing and data analysis
>* SciPy for advanced scientific computing
>* Pandas for general-purpose data analysis
>* Seaborn for (fancy) data visualization
* Platform independence and portability (Windows, IOS, Linux)
* Great community and popularity


We mentioned above that **Python is an interpreted language**. If you're new to programming, you probably have no idea what this means. Let's take a quick(ish) detour through the history of computing.

## What is programming?

What do we actually mean when we talk about coding or programming? This sounds like an obvious question, but it's worth discussing.

When we write code (in any language from Python to C++) we're writing instructions that we want our computer to execute; those instructions could be as simple as 'add these two numbers and show me the result' or as complex as 'navigate this driverless car through rush hour traffic.'

The CPU (central processing unit, or 'brain') in any computer can only understand and execute commands written in machine code, which looks something like this:

<img src="./img/machine.png" style="width:450px">

For humans, machine code is:

* Very error prone.
* A nightmare to debug: imagine trying to spot a rogue 1 (Star Wars reference entirely unintentional but I'll take it) in a string of thousands of numbers and letters so you can fix your mistake.
* Inefficient: it takes many lines of code to write even the simplest instructions.

Luckily, nowadays we have at our disposal a huge range of programming languages from JavaScipt to Python to C++. All of these are much more similar to the English language, have their own vocabulary (commands that can be given to the computer) and syntax (rules for combining commands to write complex instructions) and feel far more intuitive than machine code.

But if computers only understand machine code, how is this possible?

<img src="./img/grace.jpg" style="width:600px">

In the 1950s a US Navy Rear Admiral named Grace Hopper was well aware of this problem. She built the first compiler, a middle-man piece of software that translated instructions written in a human-readable programming language called A-0 into computer-readable machine code.

All the programming languages you might have heard of including FORTRAN, C++, Python, JavaScript, Swift, R and Matlab only exist today because of Hopper's innovation.

These modern languages are compiled or interpreted into a format that can be directly executed by a CPU. The main difference between a compiled language (like C++) and an interpreted language (like Python) lies in the result of the compilation or interpretation process.

Code written in an interpreted language is executed by an interpreter, which reads the source code and translates it on the fly into machine code, and gives us the desired result or output of the code- this could be a number, some text, or movement in a robotic arm. This means we can type statements into the interpreter and they are executed immediately, giving us near-instant results. The source code has to be re-interpreted each time the code is executed.

Lines of code written in compiled languages are converted into machine code by a compiler, to produce an executable file. To get the desired results of the code, this executable file then needs to be run. Most of the software we use on a daily basis is delivered as compiled binaries.

A program should not be confused with an *algorithm*. An algorithm is a description or sequence of steps required to  solve a problem. It is independent of any programming language or syntax,  and amounts to pseudo-code.  The steps defined in an algorithm are translated to the required language syntax, so it is the precursor to writing a program.  It means, understanding the problem we are modelling and  analysing it.  Often it will also incorporate a numerical scheme. 


***

**Interpreted** versus Compiled 1 

Programs are indirectly executed by an interpreter program which reads the source code and translates it while in motion, into a series of computations and system calls. The source has to be re-interpreted (and the interpreter present) each time the code is executed.
* Slower than compiled, with limited access to the underlying operating system and hardware
* Easier to program
* Less strict on coding errors


Interpreted versus **Compiled** 2

Most conventional kind of language is compiled. Programs are converted into machine code by a compiler and then directly executed. This executable file can be run without the need to refer back to the source code.
* Give excellent performance with limited access to the underlying operating system and hardware
*	Can be difficult to program in
*	Most of the software we use is delivered as compiled binaries, from software which the use doesn’t see


<img src="./img/Python link.png" style="width:400px">

## What sort of language is Python?
<img src="./img/scale.png" style="width:600px">

### C++ vs. Python
C++ is very fast with very optimized compilers. For intense and heavy computations, it’s hard to outperform C++. In fact it's widely agreed in banking that "C++ is used in situations where speed is everything". The language retains its status as being *sexy*! More recently, some very optimized scientific libraries have been developed for these languages, e.g. BLAS (Basic Linear Algebra Subprograms). 
There are however plenty of drawbacks! It is a very hard langauge for non-computer scientists. It's difficult to use and not easy to learn - being non-interactive during development makes its usage quite painful. Its syntax can also be verbose. Manual memory management (via C) can also be tricky.

Python may not be up there with C++ when it comes to speed, however, it makes up for this limitation in versatility.

### Distributions of Python

Python is a language, then there exist different distributions of Python (Python official, Anaconda Python, Enthought Canopy, pythonxy . . . ). Each distribution provides specific tools (Editors, packages pre-installed, support for a given platform . . . ). 


### Getting started

On Linux and Mac OS, Python comes pre-installed with the operating system. However, many useful packages (e.g. SciPy) must be installed by hand. For its easy interface we will use Anaconda. This is an open source data science platform by CONTINUUM that has many tools; these include a Python distribution, a package manager conda, a way to manage environment and many preinstalled libraries and packages (e.g. NumPy, SciPy and Jupyter notebooks). Use of Anaconda greatly facilitates the learning process and avoids many issues new programmers to Python face. More importantly it is excellent for teaching in the classroom. To download simply google Python Anaconda. You will be taken to https://www.anaconda.com/download/


<img src="./img/distribution.png" style="width:800px">

### How to know if your PC/laptop is a 32 or 64-bit system?
Linux uname -a and if you see x64 then 64-bit else 32-bit

Mac uname -a also

Windows (0ld) Click Start, right-click My Computer, and then click Properties if you see 64-bit it’s a 64-bit system

Windows 10 In search, type device specifications then System Specifications. 


Choose your operating system from Windows, mac or Linux.

<img src="./img/install.png" style="width:100x">

### Python 2 or Python 3

* Python 2.x is legacy code
* Two Python versions are in current use: Python 2 and Python 3.
* Python 3 is not backward compatible with Python 2.
* The largest community is still using Python 2 since there are still several packages incompatible with Python 3.

We will however be using a version of Python 3

### Anaconda – What is the jupyter notebook?

We've downloaded Python 3.8. What does that actually mean? What exactly have we just installed on our machines?

When we download and install Python 3.8, we download the Python 3.8 interpreter - that is, a program that can take lines of code written in Python and translate them down and show us the results. That's great! But we need somewhere to feed lines of Python code to that interpreter, or in other words we need somewhere to write, test, and run our code and view the results or output. That's what an IDE or Interactive Development Environment is for.
An IDE is anywhere that you can write and run code, and see the results. There are lots of different IDEs available- some can be used to run code in many different programming languages (like Microsoft Visual Studio Code) and some can only handle one language (like pyCharm).
***

The Jupyter Notebook (previously known as IPython) is a nice web application that allows you to create and share documents that contain live code, equations, visualizations and explanatory text. Uses include: data cleaning and transformation, numerical simulation, statistical modelling, machine learning and much, much more. Notebooks have the .ipynb extension, although it is possible to save as a .py file or other formats.
There are a number of advantages of using a Jupyter notebook. These include
* interactive
* easy to document and share results
* can access multiple kernels (languages - e.g.R)

Along the toolbar at the top you will see a tab labelled **kernel**. A notebook kernel is a “computational engine” that executes the code contained in a Notebook. The ipython kernel, executes python code, although Kernels for many other languages exist (e.g. R). Upon launching a Notebook file, the associated kernel is automatically launched. 

It is much faster to write a working Python program than a C program. The advantage of Python being a cross platform language means the same program will work immediately on almost any computer system - even on android phones!

Many researchers (and institutions) have become reliant on Jupyter notebooks to explain their results. Readers can reproduce those results and modify to create different output in order to facilitate the learning process.

For Python it is recommended to use Google Chrome and avoid Internet Explorer (some key features are disabled in Internet Explorer). However with the latter if pictures cannot be viewed, then try the following:

Normally in Internet Explorer, under Internet Options, there is an Advanced tab; under multimedia ensure "Show pictures" is enabled. 

To get started visit the following link: 
https://docs.anaconda.com/anaconda/user-guide/getting-started/

## Python script
Python script is a set of instructions in a text file, written in the Python language, that run in order and carry out a series of commands.


Begin by opening the Anaconda prompt from the start menu - this gives some technical bumf including Python version and Anaconda distribution. Note the >>> prompt. Ctrl Z or exit() both close this.  
<img src="./img/prompt.png" style="width:450x800px">

<img src="./img/navigator.png" style="width:900px">
***

If you have a Mac, then at this point click on jupyterlab
<img src="./img/jupyterlab.png" style="width:300px">
***
Under the third option (Other) will be an icon labelled **terminal**. Opening that will allow you to execute the command given below

Conda is a (open source) package manager system allowing interaction with Ananconda. It runs, installs and updates packages. Always useful to use conda before an application to check for updates, using the command **conda update python**

Typing **jupyter notebook** in Anaconda prompt opens up the Jupyter notebook. Do not close this otherwise connection to the server will be lost. 

After you run jupyter notebook your browser will open the jupyter home directory. If you are creating a new notebook, then on the right hand top corner click "new" and select "Python 3". This will open you the new jupyter notebook. After that press 'H' on your keyboard to see the list of commands. Create cell and type into it
***
If you have a notebook stored in any location on your computer, click on "upload" and add to the directory of files.

# Using Jupyter - getting started

##  Introduction

**Launching** Jupyter is usually launched from the command line using the command **jupyter notebook**. It runs in your web browser, but doesn't require an internet connection to work - remember, Python and the IDE are both running locally in your machine, your browser is just a convenient user interface.

A new notebook can be launched by selecting New, followed by Python 3.
<img src="./img/python 3.png" style="width:300px">
***
**Cells** Cells are the building blocks of Jupyter notebooks, and can contain different types of content from code, to markdown (nicely formatted text, like this cell) to LaTeX (for writing mathematical formulae) and embedded images. During this seminar we'll be using mostly code cells (to write and run our code) and markdown cells.

New cells are inserted by selecting Insert, followed by Insert Cell Above or Insert Cell Below. The dropdown menu allows us to switch between Markdown cells and Code cells.

**Running cells** Cells run one after the other. You can run cells by pressing Shift + Enter. The output from a code cell (if there is any) will appear below the cell.

**Stopping cells** running If you'd like to stop a cell running (maybe you've accidentally written some code that runs in an infinite loop) you can use the 'stop' button on the toolbar. Or go to the kernel drop-down menu and click 'Restart'; you may lose recently defined variables. 

**Saving your work** Jupyter will autosave your work once every couple of minutes, but if you've just completed a very critical piece of code, it's best to save manually as well.

**Shortcuts** Jupyter has plenty of shortcuts to save time:

<mark>Shift + Enter</mark>: Run current cell and take cursor to the next one

<mark>Ctrl + Enter</mark>: Run current cell

<mark>Alt + Enter</mark>: Run current cell and insert new cell below

<mark>Esc+M</mark>: Convert current cell to markdown

<mark>Esc+Y</mark>: Convert current cell to code

<mark>Esc + A</mark>: Insert new cell above current cell

<mark>Esc + B</mark>: Insert new cell below current cell

<mark>Esc + H</mark>: Show shortcuts menu

Notebooks have 2 modes:
1. command mode - the cell is blue (Esc key). Adding/deleting cells; using shortcuts
2. edit mode - the cell is green (Enter key).
First time you open up a notebook worth taking the user interface tour under **Help**
Using help(object) gives information on object. So e.g. if we want help on the function print we type help(print) in the cell below and run

By double clicking on a cell, you can view its content, i.e. the codde for Python or that of Markdown.

## nbextensions

Jupyter Notebook extensions or nbextensions have become an extremely popular way to extend the basic functionality of the IPython notebook ecosystem. They are simple add-ons which make your environment resemble the toolbars of a MS Office environment. They not only provide greater usage, flexibility and control;  more importantly they provide users with control to produce visually stunning notebooks. Increasing numbers of academics are now using Jupyter notebooks to produce literature with a view to sharing. In addition they aid pedagogy as they are excellent teaching tools, and the visual effects enhance the learning process.

The basic features installed are very limited. So we start by ensuring that nbextensions are available. A preferred method is to use conda

**conda install -c conda-forge jupyter_contrib_nbextensions**

The above is on the basis that the default directory is C. <br> After the installation is complete, exit the Anaconda prompt and relaunch Jupyter notebook.<br> You will now see the nbextensions tab, which looks frightfully nice!
<img src="./img/nbextensions.jpg" style="width:1050px">
<br> And this compared to the limited view you may currently have.

If the above doesn't work you can try the following three steps
* **pip install jupyter_contrib_nbextensions**
* **jupyter contrib nbextension install --user**
* **jupyter nbextension enable varInspector/main**

It is worth going through the list and reading the description for each extension. By clicking on the name 

## Running code

**Line Numbers:** To switch on (or off) line number display, use **Esc L** when in command mode

In [1]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



**Getting help - Why it's okay to Google stuff**

As mathematicians we certainly don't bother remembering dozens of results from differentiation and integration. As long as we can recall the standard important ones; there are tables of derivatives and integrals we can consult. Similarly no one has the time to memorise every single Python function they'll ever need to use (and even if they did, it would be a very inefficient use of their time). Beyond a very small handful of functions, you're not expected to memorise the commands we cover in this course, and you should never hesitate to look for help online when you get stuck.Standard sources are

* Stackoverflow: https://stackoverflow.com/
* Python documentation: https://docs.python.org/3/
* Google (obviously)

and many more.

A complete list of Python keywords is available in the **keyword** module:

In [2]:
import keyword # tests for Python keywords

Let's see the list below, some of which we will be using later

In [3]:
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

## Builtin Documentation
Jupyter notebooks are being updated with amazing speed, which means continuous inclusion of very helpful functions. One of the most useful of this platform is to shorten the learning curve by offering efficient help/documentation search. The following are the various ways to access *doc string*.

* accessing documentation with ``?``<br><br>
* accessing source code with ``??``<br><br>
* exploring modules with Tab-Completion

## Importing Modules/Libraries
A major feature of Python as an ecosystem, compared to just being a programming language, is the availability of a large number of libraries and tools. These libraries and tools generally have to be imported when needed (e.g. a plotting library) or have to be started as a separate system process (e.g., a Python development environment). Those of you familiar with (say) Matlab and R will know that they come with a pre-bundled set of modules of various aspects of scientific computing. Importing means making a library available to the current namespace and the current Python interpreter process. The reserved word import is used; just as in C++ where the prefix #include is the syntax for importing a library.  

Python itself already comes with a large set of libraries that enhance the basic interpreter in different directions. For example, basic mathematical calculations; I/O; string manipulation, can be done without any importing, while operations requiring (say) more complex mathematical functions need to be imported through the math library. When a Python program starts it only has access to a basic functions and classes.

   ("int", "dict", "len", "sum", "range", ...)
“Modules” contain additional functionality. Use the command “import” to tell Python to load a module, e.g.

In [4]:
import matplotlib # graphics library
import scipy # scientific python 
import nltk  # for natural language processing 
import numpy # numerical library and manipulating arrays - flagship
import pandas # panel data - series and dataframes - also flagship

Let's see what modules are now in the directory

In [5]:
dir()

['In',
 'NamespaceMagics',
 'Out',
 '_',
 '_3',
 '_Jupyter',
 '__',
 '___',
 '__builtin__',
 '__builtins__',
 '__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_dh',
 '_getshapeof',
 '_getsizeof',
 '_i',
 '_i1',
 '_i2',
 '_i3',
 '_i4',
 '_i5',
 '_ih',
 '_ii',
 '_iii',
 '_nms',
 '_oh',
 'exit',
 'get_ipython',
 'getsizeof',
 'json',
 'keyword',
 'matplotlib',
 'nltk',
 'np',
 'numpy',
 'pandas',
 'quit',
 'scipy',
 'var_dic_list',
 'yapf_reformat']

After using any of the modules, if we have no further use, the reserved word <span style="color:green">del</span> can be used to remove an imported module. Let's remove nltk and check the directory

In [6]:
del nltk
dir()

['In',
 'NamespaceMagics',
 'Out',
 '_',
 '_3',
 '_5',
 '_Jupyter',
 '__',
 '___',
 '__builtin__',
 '__builtins__',
 '__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_dh',
 '_getshapeof',
 '_getsizeof',
 '_i',
 '_i1',
 '_i2',
 '_i3',
 '_i4',
 '_i5',
 '_i6',
 '_ih',
 '_ii',
 '_iii',
 '_nms',
 '_oh',
 'exit',
 'get_ipython',
 'getsizeof',
 'json',
 'keyword',
 'matplotlib',
 'np',
 'numpy',
 'pandas',
 'quit',
 'scipy',
 'var_dic_list',
 'yapf_reformat']

This reserved word can also be used to delete imported functions. 

Use of object notation is encouraged. The method gives the programmer more control and the code is more readable. numpy is used as an example. We will import numpy and henceforth use the object np. So the syntax is 

In [7]:
import numpy as np

Each time in code mode you type np. you are offered the auto-complete option

In [8]:
np.__version__ # gives the current version. Note that these are double underscores pre and proceding version

'1.18.5'

For finance based work there are two useful sources (more on this later) - 

https://uk.finance.yahoo.com/

https://www.alphavantage.co/

## Magics
Notebooks have built-in special commands called **magics**. These commands are preceded by % or %%.

%: all command arguments come from that same line. These are called **line magics**.

%%: the entire cell will be used as the command arguments, hence called **cell magics**.

Let's list these available magics


In [9]:
lsmagic # this works without %

Available line magics:
%alias  %alias_magic  %autoawait  %autocall  %automagic  %autosave  %bookmark  %cd  %clear  %cls  %colors  %conda  %config  %connect_info  %copy  %ddir  %debug  %dhist  %dirs  %doctest_mode  %echo  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %macro  %magic  %matplotlib  %mkdir  %more  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %pip  %popd  %pprint  %precision  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %ren  %rep  %rerun  %reset  %reset_selective  %rmdir  %run  %save  %sc  %set_env  %store  %sx  %system  %tb  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%capture  %%cmd  %%debug  %%file  %%html  %%javascript  %%js  %%latex  %%markdown  %%perl  %%prun  %%pypy  %%python 

In [4]:
pwd # print working directory

'C:\\Users\\rahmad\\Desktop\\Fitch'

The latex below has been written in markdown
$$\alpha+\gamma$$

$$\frac{x}{y}$$

$$\int_{-\infty}^{\infty} e^{-x^2} dx = \sqrt{\pi}$$
What happens if this were done in code mode?

In [5]:
%%latex # with this cell magic all latex code goes in this cell?
$$\alpha+\gamma$$

$$\frac{x}{y}$$


$$\int_{-\infty}^{\infty} e^{-x^2} dx = \sqrt{\pi}$$

<IPython.core.display.Latex object>

There is a myriad of resources for learning latex, simply Google it. Here's a couple that sprang up

https://oeis.org/wiki/List_of_LaTeX_mathematical_symbols

http://www.docs.is.ed.ac.uk/skills/documents/3722/3722-2014.pdf

## Spell check
Given that notebooks have become an integral part of presentations, it's important that where possible typos and corrected. Although there isn't a spell check built-in, one can be imported using the following commands in Anaconda   

pip install jupyter_contrib_nbextensions

jupyter contrib nbextension install --user

jupyter nbextension enable spellchecker/main

Words not recognized by spell check are highlighted.

Python's simplest built in data types (atoms):

1. booleans
2. integers
3. floats
4. strings

* These are essentially the 'atoms' that make up the larger 'molecules'
* The object type determines what can be done with the data and whether the contents can be changed (mutable or un-mutable).
* Python is strongly typed - and object does not change.

The official definition of **strongly typed** is:
A language in which each type of data (such as integer, character, hexadecimal, packed decimal, etc.) is predefined as part of the programming language and all constants or variables defined for a given program must be described with one of the data types. 

There are a number of data structures in Python. We will introduce three basic types; **lists** e.g. ["Beijing", "Shanghai", "Chengdu] and **tuples** e.g. ("Shanghai", 7). In addition the dictionary structure will be covered. 

## Traditional First Programme
If you are familiar with the basics of programming, feel free to skip this section.

It is customary to be introduced to any new language by writing the first programme/code that simply outputs a greeting message to the screen. In e.g. C++ such a programme gives a lot of insight to the structure of a programme. So to conform with tradition we do something similar. 


In [12]:
print('hello')

hello


In [13]:
print("hello again")

hello again


What do you think will happen upon running the next cell?

In [14]:
print "hello?"

SyntaxError: Missing parentheses in call to 'print'. Did you mean print("hello?")? (<ipython-input-14-04090a97af00>, line 1)

Note single quoted and double quoted are the same – Python is one of a few programming languages where ' ' and " " have the same functionality. A general rule of thumb is to choose one and stick with it. Later we will discuss when there may be exceptions to this rule. Comments follow the octothorpe character #. A string is a sequence of characters enclosed in single or double quotes. For those familiar with version 2 of Python, you will note the insistence in version 3 with the use of brackets, as print is a function.

You must remain consistent with quote marks, i.e. you will not write ' ....... ".

Strings are of lesser importance for computational finance, although they are frequently encountered when dealing with data files, especially when importing or when formatting output. Use of strings can facilitate user friendly code.

### Python is interactive
What makes Python particularly fun is it's interactive nature. You can receive immediate feedback for for each statement, while running previously fed statements in active memory. This also allows the use of interactive mode as a calculator. Here are a few examples: $x+y$, $z^n$, $\sqrt{m}$

In [ ]:
3+2 

In [ ]:
8**3 

In [ ]:
from math import sqrt
import math as math # this is good programming
math.sqrt(16) 

The last operation has been performed without importing the math library. The following command will now produce a result

In [15]:
from math import sqrt
sqrt(16)

4.0

In [16]:
del sqrt # this will delete the square root function

Multiple statements can be expressed on the same line separated by commas

In [17]:
3+2, 8**2, math.sqrt(16)

NameError: name 'math' is not defined

What happens if you use semi-colons? Try it!

If you’ve used semicolons in C++ or Java, you will appreciate they are an integral part of their syntax. Interestingly they have no use in Python. In fact, it’s not even considered “Pythonic” to use semicolons. If we wish to give them some importance, then a semicolon may be used to separate multiple commands on a single line.

In Python groups of statements are executed one after the other:

In [ ]:
3+2; 8**3; math.sqrt(16)

In [ ]:
3+2,
8**2, 
math.sqrt(16)

In [ ]:
math.exp(1.0) # exponential function

The earlier importing of the sqrt function means that each time a different function is required, it needs to be called making the use of built-in functions quite cumbersome. The following command is a lazy alternative, for beginners to Python.
It is also an example of clumsy programming, leading to potential confusion with other libraries/you don't know where the function has come from. 

In [ ]:
from math import *

or we can adopt a short name for an imported module (which is a popular technique)

In [6]:
import math as m # now the m is an object

then use m. before each function call. For example using the exponential function would now be

In [19]:
m.exp(1.0)

2.718281828459045

In [20]:
print(m.sqrt)

<built-in function sqrt>


In [21]:
m.sqrt?

The last method is advisable and easy to read (not to mention efficient)

## The $\log$ function
The standard bases for a numeric value $x$ are logarithm 

**base 10** log10(x)

**base 2** log2(x)

**base e** log(x)

And for any general base $a$ is log(x,a). Consider the following examples


In [22]:
m.log10(4), m.log(4,10)

(0.6020599913279624, 0.6020599913279623)

In [23]:
m.log(14,5)

1.6397385131955606

In [24]:
m.log(14), m.log(14,m.e)

(2.6390573296152584, 2.6390573296152584)

In [25]:
m.pi

3.141592653589793

This is the methods approach in OOP.
The notation m.exp(1.0) is a first example of object-oriented programming (OOP). Being a mathematical function, the object m owns the method function that is called using the dot
notation. No further knowledge of OOP than understanding the notation . is required at this stage.
This differs from function(.....)

**A comprehensive list of mathematical functions can be obtained from**
https://docs.python.org/2/library/math.html

## tab and autocomplete
<img src="./img/tab.png" style="width:350px">

In [9]:
m.sin(m.pi/2)

1.0

In [10]:
m.pi

3.141592653589793

Having seen the advantage of autocomplete (using the tab key) there is another useful step. Suppose we wish to see the parameter/argument list of any built-in function e.g. the log function. Here type m.log()
Then with the cursor inside the parenthesis **shift tab** displays the (shortened) list; while doing it twice shows the complete list.

<img src="./img/auto.png" style="width:650px">

Try the above with a different function here

# Strings and operations 

A **string** is a sequence of characters, comprising of letters, numbers or symbols. In some languages a string is defined as an array of characters. We have already discussed the defining of an array using single or double speech marks. 

An empty string can also be created by a pair of single quotes or double quotes with nothing within the quotation marks.

In [11]:
str1=''
str2=""
print("string 1 is, ", str1, "i.e.empty")
print("string 2 is, ", str2, "i.e.empty")

string 1 is,   i.e.empty
string 2 is,   i.e.empty


In [13]:
str1?

Two strings can be joined/added together (concatenated) using the **+** operator:

In [30]:
"Hello, "+"Riaz!"

'Hello, Riaz!'

This joining together of strings is very simple; however if you want words split by a space you have to put the space in. Here are some examples

In [31]:
"Hello, "   +  "world!" # space after the comma

'Hello, world!'

In [32]:
"Hello,"  +  " world!" # space before second word

'Hello, world!'

In [33]:
"Hello,"  +  "world!" # no spaces included


'Hello,world!'

Text is a string of characters. Text can be converted to genuine integers using a built-in function int()

In [34]:
x=int("10")
type("10") #function type returns the data tyep

str

In [35]:
int("-100")

-100

In [36]:
int("100-10") # this gives an error

ValueError: invalid literal for int() with base 10: '100-10'

The following is allowed 

In [37]:
int('10')-int("100")

-90

## Escape sequences

Like C++, Python also has escape sequences for string literals, to insert whitespace into strings. 

Two commonly used ones are the new line code **\n** and the tab code **\t**.

In [38]:
print("Hello,\nworld") # inserts a new line

Hello,
world


In [39]:
print("Hello,\tworld") # inserts a tab

Hello,	world


In Python 3, by default the tab size is 4 chars. There are also others:

\r = carriage return (basically also used for printing stuff on the next line)

\’ = print a single quote ( ‘ ) in your text

\” = print a double quote ( “ ) in your text

\\ = print a backslash ( \ ) in your text


## String length

To find out the length of a string, use the function **len()**:

In [40]:
len("supercalifragilisticexpialidocious") # from Mary Poppins

34

In [41]:
len("supercali fragilisticexpialidocious") # whitespace is also counted

35

We can join several copies of a string with the operator *

In [42]:
3*"AbC"  # nothing exciting!

'AbCAbCAbC'

## Slicing - capturing a substring
**Slicing** is a method used to extract a portion of the string, i.e *substrings*. Slicing uses [ ] to contain the indices of the characters in a string. A string with n characters, has the first index at 0, and the last is n-1.  Here is an example:  

In [14]:
string = "hopefully bojo's brexit will fail!"
print(string[0:14]) # from 0 up to 13
len("hopefully bojo's brexit will fail!")
type(string)

hopefully bojo


str

Given a string variable s, the most commonly used types of slices are 

s[i], returns the character in position i , 

s[:i], which returns the leading characters from positions 0 to i-1, and 

s[i:], returns the trailing characters from positions i to n-1. 

s[i:j:k], returns the characters from i to j-1 in steps of k

A string can be accessed from the right of the string that is the last character on the right has the index of -1. The character before last has the index of -2, and so on.

<img src="./img/hello string.png" style="width:550px">

In the string “Hello!”, the character “H”, while accessing from left-to-right, has the index of 0 and while accessing from right-to-left has the index of -6. On the other hand While accessing the string elements from left-to-right index start with 0 and keep on incrementing by 1, whereas accessing the string elements from right-to-left the index starts with -1 and keep decrementing by 1. 

<br> <font color = black  size=4.5> Exercise 

1. Try each of the above
2. What does s[-i] return?

**Run them first before asking!!**

In [44]:
string[-2]

'l'

In [45]:
string[-34] #-34 from the end of the string equates to position zero

'h'

Staying with our variable called string, there are a number of methods available for manipulating variables of this type
string. Then hit the tab key

In [46]:
string.capitalize() # capitalises the very first letter in the string

"Hopefully bojo's brexit will fail!"

To convert each letter to upper case the method called upper.

In [47]:
string.upper()

"HOPEFULLY BOJO'S BREXIT WILL FAIL!"

To convert each element to lower case, we can use the following:

In [48]:
"WILL DR TRUMP BE A ONE TERM PRESIDENT?".lower()

'will dr trump be a one term president?'

In [17]:
substring=string.split() # use the split method and call it by another name
substring

['hopefully', "bojo's", 'brexit', 'will', 'fail!']

In [50]:
substring[0] # we can access each element of substring using indexing

'hopefully'

A string is an **immutable** object. Its individual characters cannot be modified with an assignment statement and it has a fixed length. Any attempt to be violate this property will result in TypeError


In [51]:
string[0]="H" # attempt to change h to H

TypeError: 'str' object does not support item assignment

to use quotation marks in strings, use \\. We can add arbitrary quotes by escaping them:


In [52]:
print("She said. \"He\'s coming.\"") # note \"...\'s....\"

She said. "He's coming."


\ is reserved.  \n for a newline, \t for a tabulation, \’ . . . 

## Type()
We can query the type of a value. The **type** function returns the type of any object. Combining types incorrectly returns an error. 
On a string it can be used as follows

In [53]:
type(string) # defined earlier

str

In [54]:
type("hello, mum")

str

**Note:** Variables themselves do not have a fixed type. It is only the values they refer to that have a type. This means that the type referred to by a variable can change as more statements are interpreted. Consider the following example, we use the string variable above which we see is type str

In [55]:
var1 = 2.0
type(var1)

float

In [56]:
var2 = 2
type(var2)

int

## Exercises

https://docs.python.org/2/library/string.html

Given the string "How to defend yourself against a man armed with a banana", write code to extract:

1. the following substring: "How to defend yo"
2. the first character only
3. the last character only
4. the last two characters
5. the substring "with a banana"

Convert the 'b' in banana to upper case

## Input from keyboard
Python has a built-in function for reading input. The function **input()** can be used to request information from the user via the keyboard. Here are a couple of examples

In [18]:
x=input("what is the value of x? ")
print("x= ", x)
type(x)

what is the value of x? 3
x=  3


str

In [19]:
name=input("what is your name? ")
print("Hello, ",name)
print(name[0]) # print 1st element of string

what is your name? Riaz
Hello,  Riaz
R


## Exercises

1. Create a program that assigns two integers, $a$ and $b$. Compute and display:
> * The sum of a and b
> * The difference when $b$ is subtracted from $a$
>
> * The product of $a$ and $b$
>
> * The quotient when $a$ is divided by $b$
>
>* The remainder when $a$ is divided by $b$
>
>* The result $\log_{10}a$
>
>* The result $ab$

# Data Types

When defining variable names they should only contain numbers, letters (both upper and lower), and underscores _. They must begin with a letter or an underscore - **Python is case sensitive**. As with most languages it is important that reserved words (e.g. **import** or **for**) are not used for variable names. A reserved word in Python is a word that's either

* Already the name of a built in function, like **print** or **sum** or **list**
* Already the name of an operator, like **and**, or **or**, or **not**
* Already the name of a type, like **int**

Bad things happen if you try to make a variable that has the same name as a reserved word. So good to become familiar with the basic and frequently used reserved words.

Here are some examples of legal user defined variable names

In [59]:
x = 1.0
X = 1.0 
X1 = 1.0
X1 = 1.0  # note redefinition is allowed
x1 = 1.0
big = 1.0
bigfoot = 1.0
Bigfoot = 1.0
_x = 1.0  # legal but not encouraged
x_ = 1.0  # as above

The following are illegal names for variables

In [60]:
x:=  1.0
1X = 1
X-1 = 1
for = 1

SyntaxError: invalid syntax (<ipython-input-60-45901d1005cd>, line 1)

Multiple variables can be assigned on the same line using commas

In [20]:
x=3; pi=3.14; name="Riaz"; letter='a'

In [22]:
print(x,pi,name,letter)

3 3.14 Riaz a


In [27]:
x,y,z = 3, 'hello' , 2.6

**Note:** To assign a value to a variable, we use the operator = (not to be confused with ‘equal to’). It is evaluated from right to left
name = “Riaz”
can be seen as  name     “Riaz”
and interpreted as ‘Riaz is assigned to the variable name’. So writing the script x = 3 reads as the integer 3 has been assigned to the variable. Note the type of variable x did not have to be declared prior to its initialisation. 
**Whitespace** is the name given to spaces, tabs and indents. Whitespace is also used to define blocks (in C++) using indentation. This promotes readability. For certain code it is necessary, and we will discuss later in the section on user-defined functions. Python is whitespace sensitive, so care needs to be taken in knowing when (and when not) to use it.

Python uses a highly enhanced memory allocation system which attempts to avoid allocating unnecessary memory. As a result, when one variable is assigned to another (e.g. to y = x), these will actually point to the same data in the computer’s memory. 

The function id() can be used to determine the unique identification number of a piece of data. Consider the following example


In [28]:
import math as m

In [29]:
x=m.pi
y=x
print(y)

3.141592653589793


In [31]:
print("id for x is, ",id(x))
print("id for y is, ",id(y))

id for x is,  2386128217360
id for y is,  2386128217360


In [32]:
# now change the value of x
x=1.0
print("id for x is, ",id(x))
print("id for y is, ",id(y))

id for x is,  2386174706960
id for y is,  2386128217360


In the above example, the initial assignment of y = x produced two variables with the same ID. Once x was changed, its ID changed while the ID of y did not, indicating that the data in each variable was
stored in different locations. 

## A quick note on storage
We aren't doing a computer science course. However it's not a bad idea to have some very basic appreciation of computer storage - it'll make us more responsible scientific programmers.
In computing a **bit** is short for binary digit and is the smallest unit of data in a computer. A bit has a single binary value, either 0 or 1. 
**Bytes** can then be expressed as units of bits. So, half a byte is four bits (called a nibble). Terminology does vary so in some systems, the term octet is used for an eight-bit unit instead of byte. In many systems, four eight-bit bytes or octets form a 32-bit word. Lets initialise a variable and check the number of bits. Again tab allows us to autocomplete. These methods require () at the end. Typically square brackets [] are used for indexing.

In [35]:
x=5
x.bit_length()

3

In [36]:
y=1000
y.bit_length()

10

## Computing with formulas
Having dispensed with the customary code, our first examples involve programs that evaluate mathematical formulas. The interactive nature of python means it can be used as a calculator. Consider a fundamental but very important problem in finance $$TV=PV(1+r)^N$$ where given an an initial investment $PV$ compounded at rate $r$ over $N$ years 

In [69]:
PV=100; r=0.1; N=3 # define and initialise variables 
TV=PV*(1+r)**N;  # expressing compounding 
print(TV)

133.10000000000005


Note $x^n$ is expressed in python as x**n

The above could have all been expressed on a single line to look like

In [70]:
PV=100; r=0.1; N=3; TV=PV*(1+r)**N; print(TV)

133.10000000000005


It still works. As would

In [71]:
print(100*(1+0.1)**3) # this is a complete Python program!

133.10000000000005


If interest rates are continuously compounded the earlier formula becomes
$$TV=PVe^{r(T-t)}$$
Then using the same numerical values gives 

In [72]:
import math as m
PV=100; r=0.1; T=3; t=0 # define and initialise variables 
TV=PV*m.exp(r*(T-t));  # expressing continuous compounding 
print(TV)

134.9858807576003


Rearranging the compounding formulas gives us a formula for discounting. That is, calculating the current worth $PV$ of money in the future $TV$.
$$PV=TV(1+r)^{-N}$$
$$PV=TVe^{-r(T-t)}$$


## Preliminary Remarks about Program Style
**Whitespace** is the name given to spaces, tabs and indents. Whitespace is also used to define blocks (in C++) using indentation. This promotes readability. For certain code it is necessary, and we will discuss later in the section on user-defined functions.The end of a line is unimportant in Python, as it almost completely ignores whitespace. The compounding formula script above was expressed in two ways and both gave the same result. However the lack of program comments, spaces, new lines and indentation makes this program quite unacceptable, and with larger programs becomes very difficult to read. Whilst there is much more to developing a good programming style than learning to lay out programs properly, this is nevertheless a good start. 
Be consistent with your program layout, and make sure the indentation and spacing reflects the logical structure of your program.  

It is also a good idea to choose meaningful names for variables. Remember that your code might need modification by other programmers at a later date. Even if a programmer refers to their own code (say) a year later, without proper documentation and use of meaningful variable names, they will need to spend considerable time refreshing them self on the various parts of the code. 

Unlike some languages, whitespace is significant in Python.  In Python a block is delimited by indentation. There's no brackets or parentheses to match up just indents. 

## Syntactic sugar
syntactic sugar is syntax within a programming language that is designed to make things easier to read or to express. It makes the language "sweeter" for human use: things can be expressed more clearly, more concisely, or in an alternative style that some may prefer.

https://en.wikipedia.org/wiki/Syntactic_sugar


A number can be represented by different types of variables:
* int for integer 
* long integers for integer not in the previous interval. They are stored in a more complex format. Such integers are printed out with an L at the end. The range of these long integers is only limited by the amount of available memory in the computer. Operations on long integers are slower than on normal integers!

For Simple Arithmetic Calculations – BIDMAS or BODMAS
$4+8/2$ is calculated as $4+(8/2)=8$

In [73]:
4 + 8 / 2 # spaces around the operator do not matter

8.0

In the current version of Python, integer division gives a real number. No type-casting required. 

In [74]:
20/3 , type(20/3)

(6.666666666666667, float)

If you are using version 2 of Python, simply write as 20.0/3 or 20/3.0 or type-caste as float(20)/3 to ensure a real value. Python 2.7 can use the Python 3 integer division behavior by including

In [75]:
from __future__ import division

**Integer powers** 

4 ** 3 is legal, but 4 * * 3 is not. Spaces around an operator don't matter. Spaces in the operator do!

In [76]:
4**3

64

In [77]:
4 * * 3

SyntaxError: invalid syntax (<ipython-input-77-b8b4a96dbdc5>, line 1)

**Integer remainders/Modulo:** Use % to obtain integer remainders. It operates on the absolute values

In [78]:
4%2 , 20%6 , -5%2

(0, 2, 1)

Unlike some other languages, variables can be **negated** directly e.g.

In [79]:
x=3
-x

-3

**Further Functions: int() and round()**

In [80]:
int(3.4) , int(3.9)

(3, 3)

In [81]:
round(3.4) , round(3.9)

(3, 4)

In [82]:
round(3.2399817,6) # 2 parameter function

3.239982

In programming languages generally, a variable name represents a value of a given type (int, float, etc.) stored in a fixed memory location. The value of a variable can be changed but not the variable type. This is not the case in Python where variables are *typed dynamically* as illustrated below 


In [83]:
x=3 # type int
type(x)

int

In [84]:
x=x**2.0
type(x) # type now changes

float

## The fraction module
The fractions module provides support for rational number arithmetic. It is essentially a two-tuple. A Fraction can simply be constructed from a pair of integers, from another rational number, or from a string. The syntax for importing the **fractions** library is

In [37]:
from fractions import Fraction as F

The type of the object created is type Fraction. Here are some examples

In [38]:
a=F(1,2) # initialise a fraction 
b=F(2,3)
type(a), type(b)

(Fraction(1, 2), Fraction(2, 3))

In [39]:
a+b, a*b, a/b

(Fraction(7, 6), Fraction(1, 3), Fraction(3, 4))

If we use the print() function to output the display we then get

In [88]:
print(a+b) # has a nicer output

7/6


In [89]:
F(24, -50) # produces an equivalent fraction in its simplest form

Fraction(-12, 25)

In [90]:
F(49)

Fraction(49, 1)

In [91]:
F('-4/5')

Fraction(-4, 5)

In [92]:
F('-0.325')

Fraction(-13, 40)

Predict the outcome of the following and then run 

In [93]:
x=type('-4/5')
print(x)

<class 'str'>


More examples at https://docs.python.org/2/library/fractions.html

## How large can a Python integer be?

In [94]:
4294967296**12

39402006196394479212279040100143613805079739270465446667948293404245721771497210611414266254884915640806627990306816

Python integers have unlimited range since the amount of bits used to store an integer is dynamic. So there is no limit - the only limit is machine memory. 

## Floating point
The most important (scalar) data type for numerical analysis purposes is the **float**. To input a floating data type, it is necessary to include a dot. So while 3 is an integer, 3.0 is a float. 

In [95]:
x=3; type(x)

int

In [96]:
x=3.0; type(x)

float

In [97]:
x=float(3); type(x)

float

So the integer 3 has now been converted to the float 3.0

**Hidden precision - ** Consider the following 

In [98]:
0.1

0.1

In [99]:
0.1+0.1

0.2

In [100]:
0.1+0.1+0.1

0.30000000000000004

Yes, really! If you are relying on this last decimal place, you are doing it wrong! We saw this earlier with the compounding formula example.


Switching between a long integer value and standard form is achievable using the float() function

In [101]:
x=2**127+2**65

In [102]:
x

170141183460469231768580791863303208960

In [103]:
float(x) # this now converts the above to standard form

1.7014118346046923e+38

In [104]:
float(100000000000000000000000)

1e+23

Standard form is also referred to as **scientific notation**. So $a\times{10^N}$ where $a\in{\mathbb{R}},  x\in{\mathbb{Z}}$ can be written in Python as aeN.

**Examples**
* 3e8 is $3\times{10^8}$
* 9.109e-19 is $9.109\times{10^{-19}}$

Earlier we converted text to integers. Text can also be converted to floats, using the function above, e.g.

In [105]:
float('100') , float ('10.0')

(100.0, 10.0)

We can do the opposite and convert different types to text, i.e. type str

The Python method **as_integer_ratio** returns a pair of integers whose ratio is exactly equal to the original float and with a positive denominator

In [106]:
x = 0.35
x.as_integer_ratio()

(3152519739159347, 9007199254740992)

In [107]:
str(10)

'10'

In [108]:
str('100.0')

'100.0'

So when converting between types 
* **int()** anything to an integer
* **float()** anything to float
* **str()** anything to string

In [109]:
str(3e8)

'300000000.0'

## More keyboard input

In [110]:
r=float(input("enter the circle radius - "))
area = 3.14*r**2
print("The circle area of radius ", r, " = ",area)

enter the circle radius - 10
The circle area of radius  10.0  =  314.0


Run the script above without the float function in line 1. Can you explain what is happening?

**Integer and float arithmetic**: When performing arithmetic operations, spaces around operators do not matter (ignored in the calculation). So 20 / 3 is evaluated in the same way as 20/3, etc. In the later verisons of python, type-casting is not required. This means whereas in previous versions the integer division 20/3 = 6; requiring float(20)/3 to produce the correct value, we now have

In [ ]:
20/3

## Deleting a variable
To delete the attachment we use the Python command **del**. The command
	del object
returns us to the state where the name is no longer known.
You can delete multiple names with the slightly extended syntax
	del thing1, thing2, thing3
This is equivalent to
	del thing1
	del thing2
	del thing3
but more compact.


In [111]:
x=10; print(x)

10


In [112]:
del x

In [113]:
print(x)

NameError: name 'x' is not defined

So the variable x is unkown

## Null Values
Sometimes we present "no data" or "not applicable". In Python the special value None is used, it corresponds to Null in Java (C++ has a pointer value Null). Eg.

In [114]:
value = None
print(value)

None


Note above when we fetch the value None, no result is printed out. What type is it?

In [115]:
type(value)

NoneType

We can check whether there is indeed a result or not using the **is** operator, which will return True or False:

In [116]:
value is None

True

## Polymorphism
The meaning of an operator depends on the type we are applying it to.

In [117]:
1+1

2

In [118]:
'a'+'b'

'ab'

In [119]:
"Brexit " + "is " + "for " + "fools"

'Brexit is for fools'

In [120]:
'1'+'1'

'11'

## Size of
In C++ there is an operator **sizeof(*type*)** which returns size in bytes of the object/type. 
In Python **getsizeof** function found in the sys module will give the byte size of the object. Here's a few examples:

In [121]:
import sys as sys # import the appropriate module
string = "G"
sys.getsizeof(string)

54

In [122]:
sys.getsizeof(float) # type class float

416

In [123]:
sys.getsizeof(23.657484884) # type float

24

In [124]:
sys.getsizeof(chr) # type float

72

## The **range()** function
It is tedious to write:
a = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]. There is a Python function that allows this to be done compactly. This is done using the range() function. So the earlier example can be expressed as

a = range(10)

range(n) returns the list of integers from 0 up to but not including n. 

range(m, n) returns the list of integers from m up to but not including n.

range(5, 10) gives [5, 6, 7, 8, 9]

range(m, n, i) returns a list of integers from m up to (n-1) in equal steps of +i

range(4, 16, 2) gives [4, 6, 8, 10, 12, 14]

So in short it generates a list of numbers, which is generally used to iterate over with for loops. 

In [125]:
x=range(4)
print(x, type(x))
x[0]

range(0, 4) <class 'range'>


0

Running the above script doesn't do anything exciting. We will look at loops in detail later. But for the timebeing run the following

In [126]:
for x in range(4,17,2):
    print(x)

4
6
8
10
12
14
16


How about printing in one line?

In [127]:
print(*range(4,17,2))

4 6 8 10 12 14 16


## Complex numbers (complex)

https://docs.python.org/2/library/cmath.html

Python has all the functions to manipulate complex numbers. Here $j=\sqrt{-1}$ replaces $i$; so consistent with engineers!

$z=x+iy=re^{i\theta}$


<img src='./img/complex.png' width=350>

To make use of all the complex maths library functions, import **cmath**. We now have all the functions to manipulate in $\mathbb{C}$.

In [40]:
import cmath as cm
print(cm.sqrt(-4))

2j


In [41]:
help(complex())

Help on complex object:

class complex(object)
 |  complex(real[, imag]) -> complex number
 |  
 |  Create a complex number from a real part and an optional imaginary part.
 |  This is equivalent to (real + imag*1j) where imag defaults to 0.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(...)
 |      complex.__format__() -> str
 |      
 |      Convert to a string according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(...)
 |  
 |  __gt__(self, value, /)
 |      Re

### $\mathbb{C}$ operations in Python

In [130]:
(1+1j)+(2+3j) # adding two complex numbers

(3+4j)

In [131]:
(1-2j)*(3+4j) # complex multiplication

(11-2j)

In [132]:
(1+2j)/(1-2j) # complex division

(-0.6+0.8j)

In [133]:
(2+3j)**2 # squaring a complex number

(-5+12j)

In [134]:
cm.sqrt(1j) # roots of a complex number

(0.7071067811865476+0.7071067811865475j)

There are other ways of handling complex numbers, making their manipulation more convenient and intuitive. The complex number $z=x+iy$ can be expressed as a two-tuple $z=(x,y)$, using the python function complex(x,y).

The complex numbers together with their Python expressions

${z_1}=2+3i$ is written as complex(2,3)

${z_2}=i$ is written as complex(0,1)

${z_3}=1$ is written as complex(1,0)

In [135]:
z1=complex(2,3) 
z2=complex(0,1)
print(z1+z2)

(2+4j)


In [136]:
from cmath import * # less sublime way!
print(sqrt(z2))

(0.7071067811865476+0.7071067811865475j)


The length of z is the modulus given by 
$$|z|=\sqrt{x^2+y^2}$$
The one parameter function **abs()** gives the modulus

In [137]:
abs(z1)

3.605551275463989

Calculating the real and imaginary part of $z\in\mathbb{C}$ can be achieved using the methods **real** and **imag** respectively

In [138]:
z1.real

2.0

In [139]:
z1.imag

3.0

The argument of $𝑧$, written $arg(z)$ is the angle between $𝑧(𝑥,𝑦)$ and the real axis. The one parameter function **phase()** evaluates the argument, in radians. The principal value is given, i.e $-\pi\leq\arg(z)\leq\pi$

In [140]:
phase(complex(1,1)) # 45 degrees

0.7853981633974483

Converting complex numbers from Cartesian $(x,y)$ to polar form $(r,\theta)$ is a fairly straightforward mathematical exercise. How is this done in Python? A single argument function **polar()** achieves this conversion


In [141]:
polar(0+1j)

(1.0, 1.5707963267948966)

In [142]:
polar(complex(0,1))

(1.0, 1.5707963267948966)

In [143]:
complex(4,1).real # returns the real part

4.0

In [144]:
complex(3,-2).imag # returns imaginary part

-2.0

# Exercises
There are numerous exercises in this chapter to help review the first four chapters. Consider using any of the libraries referenced earlier including https://docs.python.org/3/library/math.html  
1. Write a short program that displays your name and complete mailing address formatted in the manner that you would usually see it on the outside of an envelope. Your program should take input from the user.
2. Evaluate the following expressions: $7-15$; $3^7+16/3$; $\sqrt{12}-e^{4}$ using a line of syntax for each.
3. Write Python code to calculate the expression $\sqrt{(4x^{2}+\pi) }x^{5}\log _{4}x$, for a choice of $x>0$.
4. Calculate $\sin\left \{\dfrac{1}{x^2+1}\right\}$ for $x=3.3$.
5. The normal body temperature of a human is 37 degrees Celsius $C$. Convert this to degrees Fahrenheit $F$ using the expression $$C=\dfrac{5}{9}(F-32)$$.
6. The volume of blood in a human body is approximately 4.7 to 5.5 litres. Convert this range to both pints and gallons.
7. Create a program that inputs from the user, two integers, $a$ and $b$. Compute and display:
> * The sum of $a$ and $b$
> * The difference when $b$ is subtracted from $a$
> * The absolute value of the difference
> * The product of $a$ and $b$
>
> * The quotient when $a$ is divided by $b$
>
>* The remainder when $a$ is divided by $b$
>
>* The result $\log_{10}a$
8. Write a short program to read in a users weight in kg. Then convert this amount into pounds and stones. **Hint:** 2.2lbs$\equiv$1kg and 1 stone is 14lbs. 
9. A clinic currently measures the heights of patients in centimetres. You have been asked to assist them by writing a short Python program that reads in this information, to convert and display into feet and inches. **Hint:** One foot is 12 inches. One inch is 2.54 centimeters.
10. The volume of a cylinder can be obtained by multiplying the area of its circular base by its height. Write a program that reads the radius of the cylinder, along with its height, from the user and computes its volume. Display the result rounded to two decimal places. The volume $V$ of a circular cone, radius $r$, and vertical height (from base to vertex) $h$ is given by $V=\pi$$r^2{\dfrac{h}{3}}$
11. 180 degrees is equivalent to $\pi$ radians. Write a program to convert degrees to radians. The user should input the degrees for conversion.
12. This question involves using Python syntax to perform arithmetic on complex numbers. Let $z_{1}=3+5i$; $z_{2}=7-i$; $z_{3}=3-2i$ and $z_{4}=4+5i$. 
>* Calculate $z_{1}\pm{z_{2}}$; $z_{3}{z_{4}}$; $\dfrac{z_{3}}{z_{4}}$. What are the real and imaginary parts of $\dfrac{z_{3}}{z_{4}}$
>* Repeat the above by expressing each complex number in polar form, to validate your answers.
13. You have just opened a new savings account. The risk-free interest rate is 3.6%  per annum compounded discretely. Write a program that begins by reading the amount of money deposited into the account from the user, as well as the rate of interest. Then your program should compute and display the amount in the savings account after 1, 2, and 3 years. Display each amount so that it is rounded to 2 decimal places. Repeat this exercise for the case where the risk free rate is continuously compounded. 
14. This problem is on discounting. Calculate the current worth of £5000 in 3 years time if the interest rate is 6% continuously compounded. 